### PR TITLE
IoT Edge Marketplace

### PR DESCRIPTION
**Pull request by formulahendry**

### CODE REVIEW

File: src/marketplace/marketplace.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+    constructor(private context: vscode.ExtensionContext, private modules?: string[]) {
+        this.localServer = new LocalServer(context, modules);
+    }
+
+    public async importModule(): Promise<any> {
+        this.localServer.startServer();
+        this.panel = vscode.window.createWebviewPanel(
+            "IoT Edge Marketplace",
+            "IoT Edge Marketplace",
```
Comment (SLdragon) [inline src/marketplace/marketplace.ts]: Can we put it in Constants

File: assets/marketplace/main.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            let repository = data.containerUri;
+            let defaultTag = data.tagsOrDigests[0];
+            let splitArr = data.containerUri.split(":");
+            if (splitArr.length > 1) {
+                const tag = splitArr.pop();
+                if (data.tagsOrDigests.includes(tag)) {
+                    repository = splitArr.join(":");
+                    defaultTag = tag;
+                }
+
```
Comment (SLdragon) [inline assets/marketplace/main.js]: remove empty line

File: assets/marketplace/main.css [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+    line-height: 18px;
+}
+
+#module-detail a {
+    color: #3D7DD3;
+    text-decoration: none;
+}
+
+#sidebar {
+    width: 0%;
```
Comment (SLdragon) [inline assets/marketplace/main.css]: remove %

File: assets/marketplace/main.css [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+    background-color: #404040;
+    /* border-style: solid;
+    border-width: 5px;
+    border-color: red; */
```
Comment (SLdragon) [inline assets/marketplace/main.css]: remove

File: assets/marketplace/main.css [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+
+#module-detail .description {
+    padding-top: 5px;
+    font-size: 15px;
+}
+
+#module-detail .inputs {
+    padding-top: 10px;
+    font-size: 13px;
+
```
Comment (SLdragon) [inline assets/marketplace/main.css]: remove empty line

File: assets/marketplace/main.js (hunk 2 above)
Comment (adashen) [reply assets/marketplace/main.js]: How about move the processing data logic to localserver, and localserver returns the processed data to front?

File: src/extension.ts [hunk 6]
```diff
@@ -30,9 +31,9 @@ export function activate(context: vscode.ExtensionContext) {
     const outputChannel: vscode.OutputChannel = vscode.window.createOutputChannel(Constants.edgeDisplayName);
     Simulator.validateSimulatorUpdated(outputChannel);
     const edgeManager = new EdgeManager(context);
-
     const simulator = new Simulator(context);
     const containerManager = new ContainerManager(simulator);
+    const marketplace = new Marketplace(context);
```
Comment (adashen) [inline src/extension.ts]: When will new the marketplace? On demand or persist an instance in context?

File: assets/marketplace/main.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        searchInput: '',
+        selectedModule: {},
+        selectedTag: "",
+        moduleName: '',
+        modules: [],
+        endpoint: document.getElementById('app').getAttribute('data-endpoint'),
+        errorMessage: "",
+    },
+    created: async function () {
+        this.modules = await this.getModules();
```
Comment (adashen) [inline assets/marketplace/main.js]: Do we have specific error page? What will be shown when there is error in api call?

File: src/marketplace/marketplace.ts [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        html = html
+            .replace(/{{root}}/g, vscode.Uri.file(this.context.asAbsolutePath(".")).with({ scheme: "vscode-resource" }).toString())
+            .replace(/{{endpoint}}/g, this.localServer.getServerUri());
+        this.panel.webview.html = html;
+        this.panel.onDidDispose(() => {
+            this.panel = undefined;
+            this.localServer.stopServer();
+        });
+
+        return new Promise((resolve, reject) => {
```
Comment (adashen) [inline src/marketplace/marketplace.ts]: When will it happen, when user choose to import from market place. And import one module. But he/she doesn't close the panel, and import another again?

File: src/extension.ts (hunk 6 above)
Comment (formulahendry) [reply src/extension.ts]: Good catch. I will remove it. It is unused.

File: assets/marketplace/main.js (hunk 2 above)
Comment (formulahendry) [reply assets/marketplace/main.js]: As metadata is just a simple json file, so I fetch data directly and do some simple data process in fronted side.

File: src/marketplace/marketplace.ts (hunk 8 above)
Comment (formulahendry) [reply src/marketplace/marketplace.ts]: If user click import, the panel will close. But if use not click import, there is still chance to show many page. I will handle it to make only one page to be shown.

File: assets/marketplace/main.js (hunk 7 above)
Comment (formulahendry) [reply assets/marketplace/main.js]: Added an error page

File: src/edge/edgeManager.ts [hunk 9]
```diff
@@ -383,12 +383,13 @@ export class EdgeManager {
...
         let moduleName: string = "";
         let moduleInfo: ModuleInfo;
+        const marketplace = Marketplace.getInstance(this.context, Object.keys(modules));
         if (template === Constants.MARKETPLACE_MODULE) {
-            const marketplace = new Marketplace(this.context, Object.keys(modules));
             moduleInfo = await marketplace.importModule();
         } else {
             moduleName = Utility.getValidModuleName(await this.inputModuleName(targetModulePath, Object.keys(modules)));
             moduleInfo = await this.inputImage(moduleName, template);
+            marketplace.dispose();
```
Comment (adashen) [inline src/edge/edgeManager.ts]: why dispose here?

File: src/edge/edgeManager.ts [hunk 10]
```diff
@@ -383,12 +383,13 @@ export class EdgeManager {
         const modules = templateJson.modulesContent.$edgeAgent["properties.desired"].modules;
         let moduleName: string = "";
         let moduleInfo: ModuleInfo;
+        const marketplace = Marketplace.getInstance(this.context, Object.keys(modules));
```
Comment (adashen) [inline src/edge/edgeManager.ts]: move this line into MARKETPLACE_MODULE block?

Comment (adashen) [reply src/edge/edgeManager.ts]: Have you tested the following scenario
1. Open two edge solution in the workspace.
2. Add module for the first solution with market place. But don't continue to import
3. Add module for the second solution with marketplace. And continue to import
4. Check whether the module added to the correct target?

File: src/edge/edgeManager.ts (hunk 9 above)
Comment (formulahendry) [reply src/edge/edgeManager.ts]: Consider one edge case: User open Marketplace page but does not do anything, then he import a C# module. The template json is changed. If the Marketplace page is not closed, he could still import the a Marketplace module, but the template/modules json info has already been updated with pervious C# module.

**Pull request merged**